patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.TML.cs

```diff
@@ -181,66 +183,19 @@ internal class ModManager
 					Main.MenuUI.SetState(uiProgress);
 				}
 
-				int counter = 0;
-
-				Task.Run(() => TaskDownload(counter, uiProgress, items));
-			}
-
-			internal static void DownloadBatch(string[] workshopIds, UI.UIState returnMenu) {
-				//Set UIWorkshopDownload
-				UIWorkshopDownload uiProgress = null;		
-
-				if (!Main.dedServ) {
-					uiProgress = new UIWorkshopDownload(Interface.modPacksMenu);
-					Main.MenuUI.SetState(uiProgress);
-				}
-
-				int counter = 0;
-
-				Task.Run(() => TaskDownload(counter, uiProgress, ids: workshopIds));
+				return Task.Run(() => TaskDownload(uiProgress, items, needsReload));
 			}
 
-			private static void TaskDownload(int counter, UIWorkshopDownload uiProgress, List<ModDownloadItem> items = null, string[] ids = null) {
-				string name = "";
-				bool hasUpdate = false;
-				ModManager mod = null;
-				int endCount = 0;
-
-				if (items == null) {
-					var id = ids[counter++];
-					mod = new ModManager(new PublishedFileId_t(ulong.Parse(id)));
-
-					hasUpdate = mod.NeedsUpdate() || !mod.IsInstalled();
-					name = id;
-					endCount = ids.Length;
+			private static void TaskDownload(UIWorkshopDownload uiProgress, List<ModDownloadItem> items, bool needsReload) {
+				foreach (var item in items) {
+					var mod = new ModManager(new PublishedFileId_t(ulong.Parse(item.PublishId)), item.DisplayName);
+					mod.InnerDownload(uiProgress, item.HasUpdate);
```

Comment:
Each item has to be downloaded serially if we are to display UI that it downloaded. 
This will lead to a bunch of conflicting UIs when more than one (any mod with dependencies)

Reply:
Downloads are still performed serially. There's no parallelisation here. This is a straight refactor of repeatedly re-queueing a task with the next index in an array, to a for loop, on the same thread.